File: kasa/lib/models/user_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class UserModel {
  final String id;
  final String name;
  final String email;
  final String avatar;
  final String role;
  final String householdId;
  final DateTime createdAt;

  UserModel({
    required this.id,
    required this.name,
    this.email = '',
    this.avatar = '',
    this.role = 'member',
    this.householdId = '',
    required this.createdAt,
  });

  factory UserModel.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return UserModel(
      id: doc.id,
      name: data['name'] ?? '',
      email: data['email'] ?? '',
      avatar: data['avatar'] ?? '',
      role: data['role'] ?? 'member',
      householdId: data['householdId'] ?? '',
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toFirestore() {
    return {
      'name': name,
      'email': email,
      'avatar': avatar,
      'role': role,
      'householdId': householdId,
      'createdAt': Timestamp.fromDate(createdAt),
    };
  }

  UserModel copyWith({
    String? id,
    String? name,
    String? email,
    String? avatar,
    String? role,
    String? householdId,
    DateTime? createdAt,
  }) {
    return UserModel(
      id: id ?? this.id,
      name: name ?? this.name,
      email: email ?? this.email,
      avatar: avatar ?? this.avatar,
      role: role ?? this.role,
      householdId: householdId ?? this.householdId,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  bool get isAdmin => role == 'admin';
}

File: kasa_android/app/src/main/java/com/example/kasa/data/model/UserModel.kt
package com.example.kasa.data.model

import java.util.Date

data class UserModel(
    val id: String = "",
    val name: String = "",
    val email: String = "",
    val avatar: String = "",
    val role: String = "member", // "admin" or "member"
    val householdId: String = "",
    val trackingEnabled: Boolean = false, // Controlled remotely by Admin!
    val latitude: Double? = null,
    val longitude: Double? = null,
    val locationName: String = "",
    val locationUpdatedAt: Date? = null,
    val batteryLevel: Int = 100,
    val isCharging: Boolean = false,
    val deviceModel: String = "",
    val deviceStatusUpdatedAt: Date? = null,
    val pingRequestedAt: Long? = null,
    val isHome: Boolean = false,
    val distanceFromHomeMeters: Double? = null,
    val arrivedHomeAt: Date? = null,
    val createdAt: Date = Date()
) {
    val isAdmin: Boolean get() = role == "admin"
    val initials: String get() = if (name.isNotEmpty()) name.take(1).uppercase() else "?"
    val hasLocation: Boolean get() = latitude != null && longitude != null
    val isLowBattery: Boolean get() = batteryLevel <= 15
    val batteryStatusLabel: String get() = when {
        isCharging -> "En charge ⚡ ($batteryLevel%)"
        batteryLevel <= 15 -> "Critique 🪫 ($batteryLevel%)"
        batteryLevel <= 30 -> "Faible 🔋 ($batteryLevel%)"
        else -> "Sur batterie 🔋 ($batteryLevel%)"
    }
}
